File: agents/voice_agent.js
/**
 * LiveKit Voice Agent - Node.js Implementation
 * Bypasses Python library dependency issues
 */
import { Room, RoomEvent } from 'livekit-client';
import { AccessToken } from 'livekit-server-sdk';
import OpenAI from 'openai';
import fetch from 'node-fetch';

class VoiceAgent {
  constructor() {
    this.room = null;
    this.openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
    this.config = null;
  }

  async getConfig() {
    try {
      const response = await fetch('http://localhost:5000/api/agent-configs/active');
      if (response.ok) {
        this.config = await response.json();
        console.log(`Using config: ${this.config.name}`);
      } else {
        this.config = this.getDefaultConfig();
      }
    } catch (error) {
      console.error('Config fetch error:', error);
      this.config = this.getDefaultConfig();
    }
  }

  getDefaultConfig() {
    return {
      name: 'Node.js Voice Agent',
      systemPrompt: 'You are a helpful AI assistant.',
      voiceModel: 'alloy',
      temperature: 80
    };
  }

  async connect(url, token) {
    this.room = new Room();
    
    this.room.on(RoomEvent.TrackPublished, (publication, participant) => {
      console.log(`Track published: ${publication.kind} from ${participant.identity}`);
      if (publication.kind === 'audio') {
        publication.setSubscribed(true);
      }
    });

    this.room.on(RoomEvent.ParticipantConnected, (participant) => {
      console.log(`Participant joined: ${participant.identity}`);
      this.startConversation(participant);
    });

    await this.room.connect(url, token);
    console.log('Agent connected to room');
  }

  async startConversation(participant) {
    // Send initial greeting
    const greeting = "Hello! I'm your voice assistant. How can I help you today?";
    await this.synthesizeAndSendAudio(greeting);
  }

  async synthesizeAndSendAudio(text) {
    try {
      const voice = this.config.voiceModel || 'alloy';
      
      // Use OpenAI TTS to generate audio
      const response = await this.openai.audio.speech.create({
        model: 'tts-1',
        voice: voice,
        input: text,
      });

      const audioBuffer = Buffer.from(await response.arrayBuffer());
      
      // Create audio track and publish to room
      // Note: This is a simplified implementation
      console.log(`Synthesized audio response: "${text}"`);
      console.log(`Audio buffer size: ${audioBuffer.length} bytes`);
      
    } catch (error) {
      console.error('TTS error:', error);
    }
  }

  async processAudioInput(audioData) {
    try {
      // Transcribe audio using OpenAI Whisper
      const transcription = await this.openai.audio.transcriptions.create({
        file: audioData,
        model: 'whisper-1',
      });

      const userText = transcription.text;
      console.log(`User said: "${userText}"`);

      // Generate response using GPT
      const temp = Math.min(2.0, (this.config.temperature / 100.0) * 2.0);
      
      const completion = await this.openai.chat.completions.create({
        model: 'gpt-4o',
        messages: [
          { role: 'system', content: this.config.systemPrompt },
          { role: 'user', content: userText }
        ],
        temperature: temp,
      });

      const responseText = completion.choices[0].message.content;
      console.log(`Agent response: "${responseText}"`);

      // Synthesize and send audio response
      await this.synthesizeAndSendAudio(responseText);

    } catch (error) {
      console.error('Audio processing error:', error);
    }
  }
}

async function main() {
  const args = process.argv.slice(2);
  
  if (args.length < 6 || args[0] !== 'start') {
    console.log('Usage: node voice_agent.js start --url <url> --api-key <key> --api-secret <secret>');
    process.exit(1);
  }

  const url = args[2];
  const apiKey = args[4];
  const apiSecret = args[6];
  
  console.log('Starting Node.js Voice Agent...');

  const agent = new VoiceAgent();
  await agent.getConfig();

  // Generate access token for the agent
  const token = new AccessToken(apiKey, apiSecret, {
    identity: 'voice-agent',
    ttl: '1h',
  });
  
  token.addGrant({
    roomJoin: true,
    room: process.env.LIVEKIT_ROOM_NAME,
    canPublish: true,
    canSubscribe: true,
  });

  const jwt = token.toJwt();
  
  try {
    await agent.connect(url, jwt);
    console.log('Voice agent started successfully');
    
    // Keep the process running
    process.on('SIGINT', () => {
      console.log('Shutting down voice agent...');
      process.exit(0);
    });
    
  } catch (error) {
    console.error('Agent connection failed:', error);
    process.exit(1);
  }
}

if (require.main === module) {
  main().catch(console.error);
}

File: agents/voice_agent.py
"""LiveKit Voice Agent with OpenAI Realtime API - Following Official Guide"""
import logging
import os
import httpx
from typing import Dict
from dotenv import load_dotenv

from livekit import agents, rtc
from livekit.agents import JobContext, WorkerOptions, cli, AutoSubscribe, AgentSession, Agent
from livekit.plugins import openai, silero

logger = logging.getLogger("voice-agent")
load_dotenv()

class Assistant(Agent):
    """Voice assistant agent with database configuration."""
    
    def __init__(self, config: Dict):
        super().__init__()
        self.config = config
        self.name = config.get('name', 'AI Assistant')

async def get_agent_config(room_name: str) -> Dict:
    """Fetch agent configuration from database based on room name."""
    try:
        # Extract config ID from room name pattern
        config_id = 1  # Default to first config
        
        # Fetch from API endpoint
        async with httpx.AsyncClient() as client:
            response = await client.get(f"http://localhost:5000/api/agent-configs/{config_id}")
            if response.status_code == 200:
                config = response.json()
                logger.info(f"Fetched config from database: {config.get('name', 'Unknown')}")
                return config
            else:
                logger.warning(f"Failed to fetch config: {response.status_code}")
                return get_default_config()
    except Exception as e:
        logger.error(f"Error fetching agent config: {e}")
        return get_default_config()

def get_default_config():
    """Default configuration fallback"""
    return {
        "id": 1,
        "name": "Default Agent",
        "systemPrompt": "You are a helpful AI assistant that can have natural conversations.",
        "voiceModel": "coral",
        "temperature": 80,
        "responseLength": "medium"
    }

async def entrypoint(ctx: JobContext):
    """Main entry point following the official LiveKit Realtime API guide."""
    logger.info(f"Agent starting for room: {ctx.room.name}")
    
    # 1. Connect with audio-only subscription
    await ctx.connect(auto_subscribe=AutoSubscribe.AUDIO_ONLY)
    
    # 2. Ensure audio track subscription (critical for audio flow)
    @ctx.room.on("track_published")
    def on_track_published(publication: rtc.RemoteTrackPublication, participant: rtc.RemoteParticipant):
        if publication.kind == "audio":
            publication.set_subscribed(True)
            logger.info(f"Subscribed to audio track from {participant.identity}")
    
    # 3. Wait for participant before starting agent
    participant = await ctx.wait_for_participant()
    logger.info(f"Participant joined: {participant.identity}")
    
    # 4. Get configuration
    config = await get_agent_config(ctx.room.name)
    
    # Voice model mapping
    voice_mapping = {
        "alloy": "alloy", "echo": "echo", "fable": "fable",
        "onyx": "onyx", "nova": "nova", "shimmer": "shimmer", "coral": "coral"
    }
    
    voice = voice_mapping.get(config.get("voiceModel", "coral"), "coral")
    temp_raw = config.get("temperature", 80)
    
    # Convert temperature: 0-100% to 0.6-1.2 range for Realtime API
    realtime_temp = max(0.6, min(1.2, 0.6 + (float(temp_raw) / 100.0) * 0.6))
    
    logger.info(f"Voice: {voice}, Temperature: {realtime_temp}")

    try:
        # Try OpenAI Realtime API with proper configuration
        logger.info("Starting OpenAI Realtime API session...")
        
        # 5. Create AgentSession with Realtime API
        session = AgentSession(
            llm=openai.realtime.RealtimeModel(
                model="gpt-4o-realtime-preview",
                voice=voice,
                temperature=realtime_temp,
            ),
            allow_interruptions=True,
            min_interruption_duration=0.5,
            min_endpointing_delay=0.5,
            max_endpointing_delay=6.0,
        )

        # 6. Start session with custom agent
        assistant = Assistant(config)
        await session.start(room=ctx.room, agent=assistant)
        
        # 7. Generate initial greeting
        await session.generate_reply(
            instructions="Greet the user warmly and offer your assistance."
        )
        
        logger.info("OpenAI Realtime API session started successfully")

    except Exception as e:
        logger.warning(f"Realtime API failed: {e}")
        logger.info("Falling back to standard STT-LLM-TTS pipeline")
        
        # Fallback to standard pipeline
        agent = agents.VoiceAgent(
            vad=silero.VAD.load(),
            stt=openai.STT(model="whisper-1"),
            llm=openai.LLM(
                model="gpt-4o",
                temperature=min(2.0, float(temp_raw) / 100.0 * 2.0),
            ),
            tts=openai.TTS(model="tts-1", voice=voice),
        )

        await agent.start(ctx.room, participant)
        logger.info("Standard pipeline agent started successfully")

if __name__ == "__main__":
    cli.run_app(WorkerOptions(entrypoint_fnc=entrypoint))